### ISSUE TITLE
Add ability to skip theories without data rather than fail them

### ISSUE THREAD

**Issue by cesarblum** [xunit/xunit#1113]
Repro code:

```csharp
using Xunit;

namespace EmptyTheoryData
{
    public class UnitTests
    {
        [Theory]
        [MemberData(nameof(TestData))]
        public void Test(int arg)
        {
        }

        public static TheoryData<int> TestData
        {
            get
            {
                return new TheoryData<int>();
            }
        }
    }
}
```

Test run fails with

```
Starting test execution, please wait...
Failed   EmptyTheoryData.UnitTests.Test
Error Message:
 System.InvalidOperationException : No data found for EmptyTheoryData.UnitTests.Test


Total tests: 1. Passed: 0. Failed: 1. Skipped: 0.
Test Run Failed.
```

This is a hassle when theory data is dynamically built based on conditions which might not be fulfilled, resulting in an empty collection. We ran into that in Kestrel in this test:

https://github.com/aspnet/KestrelHttpServer/blob/1a2c4388991c0583909f9c9d6e2f6aea48efa915/test/Microsoft.AspNetCore.Server.Kestrel.FunctionalTests/AddressRegistrationTests.cs#L73

There are machines where there are not IPv6 addresses with scope IDs, and the test run fails on those. This fails our test runs on AppVeyor.

I don't see why an empty collection should fail the test run. Skipping the test in that case would be better.

**Comment by bradwilson** [comment]
We have long considered a theory with no data to be an error condition (it's been like this for many years). Making this change would cause people to be able to inadvertently write a test they think runs, but actually doesn't because there's no data. This is desired behavior that we intend to preserve.

It sounds like what you should be looking for, instead, of a dynamically skip-able theory (skip it environmentally, because IPv6 is not enabled), or a theory extension which behaves differently in the absence of data.

**Comment by vbreuss** [comment]
@bradwilson I also repeatedly stumbled upon this issue, and I think it would be a great addition to V3 to add an optional parameter (e.g. `AllowEmptyData = true`) to `TheoryAttribute` so that the test is successful or skipped in case of empty data.

**Comment by bradwilson** [comment]
I've re-opened it. I'm okay with this not being the default behavior but something that can be opted into.

**Comment by bradwilson** [comment]
Available in v3 `0.2.0-pre.66` https://xunit.net/docs/using-ci-builds

**Comment by bradwilson** [comment]
I found a bug where dataless tests wouldn't be failed _or_ skipped if theories weren't pre-enumerated.

The bug fix for that is in `0.2.0-pre.67`.
